### PR TITLE
[IOTDB-3433] add metric for cluster.

### PR DESCRIPTION
**Pull request by SpriCoder**
See more details: https://issues.apache.org/jira/browse/IOTDB-3433

### CODE REVIEW

**Comment by SpriCoder** [comment]
Get it, I will add some related metrics.

**Review by xingtanzjr — APPROVED** [review]

**Comment by SpriCoder** [comment]
In Grafana, for confignode:
image

for datanode:
image

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java [hunk 1]
```diff
@@ -98,6 +105,42 @@ public PartitionInfo() {
     this.deletedRegionSet = Collections.synchronizedSet(new HashSet<>());
   }
 
+  public void addMetrics() {
+    if (MetricConfigDescriptor.getInstance().getMetricConfig().getEnableMetric()) {
+      MetricsService.getInstance()
+          .getMetricManager()
+          .getOrCreateAutoGauge(
+              Metric.PARTITION_TABLE.toString(),
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:113]: Metric.STORAGEGROUP?

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java [hunk 2]
```diff
@@ -98,6 +105,42 @@ public PartitionInfo() {
...
+              MetricLevel.CORE,
+              storageGroupPartitionTables,
+              Map::size,
+              Tag.NAME.toString(),
+              "number");
+      MetricsService.getInstance()
+          .getMetricManager()
+          .getOrCreateAutoGauge(
+              Metric.REGION.toString(),
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:122]: Metric.SCHEMAREGION?

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java [hunk 3]
```diff
@@ -98,6 +105,42 @@ public PartitionInfo() {
...
+              this,
+              o -> o.getTotalRegionCountAndUpdateMetric(TConsensusGroupType.SchemaRegion),
+              Tag.NAME.toString(),
+              "total",
+              Tag.TYPE.toString(),
+              TConsensusGroupType.SchemaRegion.toString());
+      MetricsService.getInstance()
+          .getMetricManager()
+          .getOrCreateAutoGauge(
+              Metric.REGION.toString(),
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:133]: Metric.DATAREGION?

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java [hunk 4]
```diff
@@ -484,6 +529,54 @@ public List<Pair<Long, TConsensusGroupId>> getSortedRegionSlotsCounter(
     return storageGroupPartitionTables.get(storageGroup).getSortedRegionSlotsCounter(type);
   }
 
+  /**
+   * Get total region number and update metric
+   *
+   * @param type SchemaRegion or DataRegion
+   * @return the number of SchemaRegion or DataRegion
+   */
+  private int getTotalRegionCountAndUpdateMetric(TConsensusGroupType type) {
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:538]: Split this interface into getTotalRegionCount and UpdateMetric might better, I wanna use getTotalRegionCount later~

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java [hunk 5]
```diff
@@ -484,6 +529,54 @@ public List<Pair<Long, TConsensusGroupId>> getSortedRegionSlotsCounter(
...
+    Set<RegionGroup> regionGroups = new HashSet<>();
+    for (Map.Entry<String, StorageGroupPartitionTable> entry :
+        storageGroupPartitionTables.entrySet()) {
+      regionGroups.addAll(entry.getValue().getRegion(type));
+    }
+    int result = regionGroups.size();
+    // statistic
+    Map<TDataNodeLocation, Integer> dataNodeLocationIntegerMap = new HashMap<>();
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:546]: Add java doc to display what this Map stands for

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java [hunk 6]
```diff
@@ -484,6 +529,54 @@ public List<Pair<Long, TConsensusGroupId>> getSortedRegionSlotsCounter(
...
+      MetricsService.getInstance()
+          .getMetricManager()
+          .getOrCreateGauge(
+              Metric.REGION.toString(),
+              MetricLevel.IMPORTANT,
+              Tag.NAME.toString(),
+              name,
+              Tag.TYPE.toString(),
+              type.toString())
+          .set(dataNodeLocationIntegerMap.get(dataNodeLocation));
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:575]: Comments may require another screenshot to show this monitoring indicators

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java [hunk 7]
```diff
@@ -70,6 +81,57 @@ public StorageGroupPartitionTable() {
...
+
+    addMetrics();
+  }
+
+  private void addMetrics() {
+    if (MetricConfigDescriptor.getInstance().getMetricConfig().getEnableMetric()) {
+      MetricsService.getInstance()
+          .getMetricManager()
+          .getOrCreateAutoGauge(
+              Metric.REGION.toString(),
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java:93]: Metric.SCHEMAREGION?

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java [hunk 8]
```diff
@@ -70,6 +81,57 @@ public StorageGroupPartitionTable() {
...
+              this,
+              o -> o.getRegionCount(TConsensusGroupType.SchemaRegion),
+              Tag.NAME.toString(),
+              storageGroupName,
+              Tag.TYPE.toString(),
+              TConsensusGroupType.SchemaRegion.toString());
+      MetricsService.getInstance()
+          .getMetricManager()
+          .getOrCreateAutoGauge(
+              Metric.REGION.toString(),
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java:104]: Metric.DATAREGION?

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java [hunk 9]
```diff
@@ -70,6 +81,57 @@ public StorageGroupPartitionTable() {
...
+          .getOrCreateAutoGauge(
+              Metric.SLOT.toString(),
+              MetricLevel.NORMAL,
+              dataPartitionTable,
+              o -> o.getDataPartitionMap().size(),
+              Tag.NAME.toString(),
+              storageGroupName,
+              Tag.TYPE.toString(),
+              "dataSlotNumber");
+    }
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java:134]: The name "SLOT" will be update later, please add a "TODO"

**Review by CRZbulabula — COMMENTED** [review]
PTAL~

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java (hunk 1 above)
Comment (SpriCoder) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:113]: Ok, I will update.

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java (hunk 2 above)
Comment (SpriCoder) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:122]: I use type tag to filter which is more convenience.

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java (hunk 3 above)
Comment (SpriCoder) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:133]: I use type tag to filter which is more convenience.

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java (hunk 4 above)
Comment (SpriCoder) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:538]: Get it.

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java (hunk 5 above)
Comment (SpriCoder) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:546]: get

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java (hunk 6 above)
Comment (SpriCoder) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:575]: this is shown in region on node.

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java (hunk 7 above)
Comment (SpriCoder) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java:93]: I use type tag to filter which is more convenience.

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java (hunk 8 above)
Comment (SpriCoder) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java:104]: I use type tag to filter which is more convenience.

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java (hunk 9 above)
Comment (SpriCoder) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java:134]: get

**Review by CRZbulabula — APPROVED** [review]
LGTM~

**Pull request merged**